In [1]:
!pip install PyPDF2 faiss-cpu torch transformers gpt4all

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 232.6/232.6 kB 5.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.5/27.5 MB 41.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.6/121.6 MB 6.4 MB/s eta 0:00:00


Collecting PyPDF2
  Downloading pypdf2-3.0.1-py3-none-any.whl.metadata (6.8 kB)
Collecting faiss-cpu
  Downloading faiss_cpu-1.9.0.post1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.4 kB)
Requirement already satisfied: torch in /usr/local/lib/python3.11/dist-packages (2.5.1+cu121)
Requirement already satisfied: transformers in /usr/local/lib/python3.11/dist-packages (4.47.1)
Collecting gpt4all
  Downloading gpt4all-2.8.2-py3-none-manylinux1_x86_64.whl.metadata (4.8 kB)
Requirement already satisfied: numpy<3.0,>=1.25.0 in /usr/local/lib/python3.11/dist-packages (from faiss-cpu) (1.26.4)
Requirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from faiss-cpu) (24.2)
Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from torch) (3.16.1)
Requirement already satisfied: typing-extensions>=4.8.0 in /usr/local/lib/python3.11/dist-packages (from torch) (4.12.2)
Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch) (3.4.2)
Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch) (3.1.5)
Requirement already satisfied: fsspec in /usr/local/lib/python3.11/dist-packages (from torch) (2024.10.0)
Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch) (12.1.105)
Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch) (12.1.105)
Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch) (12.1.105)
Requirement already satisfied: nvidia-cudnn-cu12==9.1.0.70 in /usr/local/lib/python3.11/dist-packages (from torch) (9.1.0.70)
Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /usr/local/lib/python3.11/dist-packages (from torch) (12.1.3.1)
Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /usr/local/lib/python3.11/dist-packages (from torch) (11.0.2.54)
Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /usr/local/lib/python3.11/dist-packages (from torch) (10.3.2.106)
Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /usr/local/lib/python3.11/dist-packages (from torch) (11.4.5.107)
Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /usr/local/lib/python3.11/dist-packages (from torch) (12.1.0.106)
Requirement already satisfied: nvidia-nccl-cu12==2.21.5 in /usr/local/lib/python3.11/dist-packages (from torch) (2.21.5)
Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /usr/local/lib/python3.11/dist-packages (from torch) (12.1.105)
Requirement already satisfied: triton==3.1.0 in /usr/local/lib/python3.11/dist-packages (from torch) (3.1.0)
Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.11/dist-packages (from torch) (1.13.1)
Requirement already satisfied: nvidia-nvjitlink-cu12 in /usr/local/lib/python3.11/dist-packages (from nvidia-cusolver-cu12==11.4.5.107->torch) (12.6.85)
Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy==1.13.1->torch) (1.3.0)
Requirement already satisfied: huggingface-hub<1.0,>=0.24.0 in /usr/local/lib/python3.11/dist-packages (from transformers) (0.27.1)
Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.11/dist-packages (from transformers) (6.0.2)
Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.11/dist-packages (from transformers) (2024.11.6)
Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from transformers) (2.32.3)
Requirement already satisfied: tokenizers<0.22,>=0.21 in /usr/local/lib/python3.11/dist-packages (from transformers) (0.21.0)
Requirement already satisfied: safetensors>=0.4.1 in /usr/local/lib/python3.11/dist-packages (from transformers) (0.5.2)
Requirement already satisfied: tqdm>=4.27 in /usr/local/lib/python3.11/dist-packages (from transformers) (4.67.1)
Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch) (3.0.2)
Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (3.4.1)
Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (3.10)
Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (2.3.0)
Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (2024.12.14)
Downloading pypdf2-3.0.1-py3-none-any.whl (232 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 232.6/232.6 kB 5.3 MB/s eta 0:00:00
Downloading faiss_cpu-1.9.0.post1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (27.5 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.5/27.5 MB 41.9 MB/s eta 0:00:00
Downloading gpt4all-2.8.2-py3-none-manylinux1_x86_64.whl (121.6 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.6/121.6 MB 6.4 MB/s eta 0:00:00
Installing collected packages: PyPDF2, faiss-cpu, gpt4all
Successfully installed PyPDF2-3.0.1 faiss-cpu-1.9.0.post1 gpt4all-2.8.2

In [2]:
import os
import numpy as np
import faiss
import torch
from transformers import AutoTokenizer, AutoModel
from PyPDF2 import PdfReader
from gpt4all import GPT4All

# تحميل النموذج اللغوي
gpt4all_model = GPT4All("mistral-7b-openorca.Q4_0.gguf")

# تحميل نموذج التضمين
tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-small-en-v1.5")
embedding_model = AutoModel.from_pretrained("BAAI/bge-small-en-v1.5")

# قراءة ملف PDF
def read_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

# تقسيم النص إلى أجزاء
def split_text(text, chunk_size=500, overlap=50):
    words = text.split()
    chunks = []
    for i in range(0, len(words), chunk_size - overlap):
        chunk = " ".join(words[i:i + chunk_size])
        chunks.append(chunk)
    return chunks

# تضمين النصوص
def embed_texts(texts):
    inputs = tokenizer(texts, return_tensors="pt", padding=True, truncation=True)
    with torch.no_grad():
        outputs = embedding_model(**inputs)
    embeddings = outputs.last_hidden_state.mean(dim=1).numpy()
    return embeddings

# إنشاء FAISS index
def create_faiss_index(embeddings):
    dimension = embeddings.shape[1]
    index = faiss.IndexFlatL2(dimension)
    index.add(embeddings)
    return index

# البحث عن النصوص المشابهة
def search_similar_texts(query_embedding, index, chunks, k=5):
    distances, indices = index.search(query_embedding, k)
    similar_texts = [chunks[i] for i in indices[0]]
    return similar_texts

# توليد الإجابة باستخدام GPT4All
def generate_answer(query, context):
    prompt = f"Context: {context}\n\nQuestion: {query}\n\nAnswer:"
    response = gpt4all_model.generate(prompt, max_tokens=100)
    return response

# المسار إلى ملف PDF
pdf_path = "/content/The_Lightning_Thief_-_Percy_Jackson_1-10.pdf"

# قراءة النص من PDF
text = read_pdf(pdf_path)

# تقسيم النص إلى أجزاء
chunks = split_text(text)

# تضمين الأجزاء
embeddings = embed_texts(chunks)

# إنشاء FAISS index
index = create_faiss_index(embeddings)

# السؤال
query = "Who is Percy?"

# تضمين السؤال
query_embedding = embed_texts([query])

# البحث عن النصوص المشابهة
similar_texts = search_similar_texts(query_embedding, index, chunks)

# توليد الإجابة
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)

Downloading: 100%|██████████| 4.11G/4.11G [01:12<00:00, 56.9MiB/s]
/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/366 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/711k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/743 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/133M [00:00<?, ?B/s]

Answer: ERROR: The prompt size exceeds the context window size and cannot be processed.


In [ ]:
import os
import numpy as np
import faiss
import torch
from transformers import AutoTokenizer, AutoModel
from PyPDF2 import PdfReader
from gpt4all import GPT4All

# تحميل النموذج اللغوي
gpt4all_model = GPT4All("mistral-7b-openorca.Q4_0.gguf")

# تحميل نموذج التضمين
tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-small-en-v1.5")
embedding_model = AutoModel.from_pretrained("BAAI/bge-small-en-v1.5")

# قراءة ملف PDF
def read_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

# تقسيم النص إلى أجزاء
def split_text(text, chunk_size=500, overlap=50):
    words = text.split()
    chunks = []
    for i in range(0, len(words), chunk_size - overlap):
        chunk = " ".join(words[i:i + chunk_size])
        chunks.append(chunk)
    return chunks

# تضمين النصوص
def embed_texts(texts):
    inputs = tokenizer(texts, return_tensors="pt", padding=True, truncation=True)
    with torch.no_grad():
        outputs = embedding_model(**inputs)
    embeddings = outputs.last_hidden_state.mean(dim=1).numpy()
    return embeddings

# إنشاء FAISS index
def create_faiss_index(embeddings):
    dimension = embeddings.shape[1]
    index = faiss.IndexFlatL2(dimension)
    index.add(embeddings)
    return index

# البحث عن النصوص المشابهة
def search_similar_texts(query_embedding, index, chunks, k=5):
    distances, indices = index.search(query_embedding, k)
    similar_texts = [chunks[i] for i in indices[0]]
    return similar_texts

# توليد الإجابة باستخدام GPT4All
def generate_answer(query, context):
    prompt = f"Context: {context}\n\nQuestion: {query}\n\nAnswer:"
    response = gpt4all_model.generate(prompt, max_tokens=100)
    return response

# المسار إلى ملف PDF
pdf_path = "/content/The_Lightning_Thief_-_Percy_Jackson_1-10.pdf"

# قراءة النص من PDF
text = read_pdf(pdf_path)

# تقسيم النص إلى أجزاء
chunks = split_text(text)

# تضمين الأجزاء
embeddings = embed_texts(chunks)

# إنشاء FAISS index
index = create_faiss_index(embeddings)

# السؤال
query = "Who is Percy?"

# تضمين السؤال
query_embedding = embed_texts([query])

# البحث عن النصوص المشابهة
similar_texts = search_similar_texts(query_embedding, index, chunks)

# توليد الإجابة
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)

Successfully installed PyPDF2-3.0.1 faiss-cpu-1.9.0.post1 gpt4all-2.8.2

شغال والاجابة رائعة

In [3]:
import os
import numpy as np
import faiss
import torch
from transformers import AutoTokenizer, AutoModel
from PyPDF2 import PdfReader
from gpt4all import GPT4All

# Load GPT4All model
gpt4all_model = GPT4All("mistral-7b-openorca.Q4_0.gguf")

# Load embedding model
tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-small-en-v1.5")
embedding_model = AutoModel.from_pretrained("BAAI/bge-small-en-v1.5")

# Read PDF file
def read_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

# Split text into smaller chunks
def split_text(text, chunk_size=300, overlap=50):
    words = text.split()
    chunks = []
    for i in range(0, len(words), chunk_size - overlap):
        chunk = " ".join(words[i:i + chunk_size])
        chunks.append(chunk)
    return chunks

# Embed texts using BAAI/bge-small-en-v1.5
def embed_texts(texts):
    inputs = tokenizer(texts, return_tensors="pt", padding=True, truncation=True, max_length=512)
    with torch.no_grad():
        outputs = embedding_model(**inputs)
    embeddings = outputs.last_hidden_state.mean(dim=1).numpy()
    return embeddings

# Create FAISS index
def create_faiss_index(embeddings):
    dimension = embeddings.shape[1]
    index = faiss.IndexFlatL2(dimension)
    index.add(embeddings)
    return index

# Search for similar texts
def search_similar_texts(query_embedding, index, chunks, k=3):
    distances, indices = index.search(query_embedding, k)
    similar_texts = [chunks[i] for i in indices[0]]
    return similar_texts

# Generate answer using GPT4All
def generate_answer(query, context, max_tokens=100):
    # Truncate context to fit within the model's token limit
    max_context_tokens = 2048 - len(tokenizer.encode(query)) - 10  # Reserve tokens for query and padding
    context_tokens = tokenizer.encode(context, max_length=max_context_tokens, truncation=True)
    truncated_context = tokenizer.decode(context_tokens, skip_special_tokens=True)

    # Create prompt
    prompt = f"Context: {truncated_context}\n\nQuestion: {query}\n\nAnswer:"

    # Generate response
    response = gpt4all_model.generate(prompt, max_tokens=max_tokens)
    return response

# Path to PDF file
pdf_path = "/content/The_Lightning_Thief_-_Percy_Jackson_1-10.pdf"

# Read text from PDF
text = read_pdf(pdf_path)

# Split text into chunks
chunks = split_text(text, chunk_size=300, overlap=50)

# Embed chunks
embeddings = embed_texts(chunks)

# Create FAISS index
index = create_faiss_index(embeddings)

# Query
query = "Who is Percy?"

# Embed query
query_embedding = embed_texts([query])

# Search for similar texts
similar_texts = search_similar_texts(query_embedding, index, chunks, k=3)

# Generate answer
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)

Answer:  Percy Jackson is a demigod in Rick Riordan's book series of the same name. He is the son of Poseidon, the Greek god of the sea, and his mortal mother, Sally Jackson. As a demigod, he possesses both human and divine traits, including superhuman strength, speed, and various other abilities related to water and sea creatures. Percy attends Camp Half-Blood, a summer camp for demigods


Answer:  Percy Jackson is a demigod in Rick Riordan's book series of the same name. He is the son of Poseidon, the Greek god of the sea, and his mortal mother, Sally Jackson. As a demigod, he possesses both human and divine traits, including superhuman strength, speed, and various other abilities related to water and sea creatures. Percy attends Camp Half-Blood, a summer camp for demigod

شغال رائع واجابة جيدة

In [2]:
import numpy as np
import faiss
import torch
from transformers import AutoTokenizer, AutoModel
from PyPDF2 import PdfReader
from gpt4all import GPT4All
from nltk.tokenize import sent_tokenize
import nltk

# Download the required NLTK resources
nltk.download('punkt')
nltk.download('punkt_tab')

# Load GPT4All model
gpt4all_model = GPT4All("mistral-7b-openorca.Q4_0.gguf")

# Load embedding model
tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-small-en-v1.5")
embedding_model = AutoModel.from_pretrained("BAAI/bge-small-en-v1.5")

# Read PDF file
def read_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

# Split text into sentence-based chunks
def split_text(text, max_chunk_size=200):  # Reduced chunk size to increase the number of chunks
    sentences = sent_tokenize(text)
    chunks = []
    current_chunk = ""
    for sentence in sentences:
        if len(current_chunk) + len(sentence) <= max_chunk_size:
            current_chunk += " " + sentence
        else:
            chunks.append(current_chunk.strip())
            current_chunk = sentence
    if current_chunk:
        chunks.append(current_chunk.strip())
    return chunks

# Embed texts in batches
def embed_texts_in_batches(texts, batch_size=32):
    embeddings = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tokenizer(batch, return_tensors="pt", padding=True, truncation=True, max_length=512)
        with torch.no_grad():
            outputs = embedding_model(**inputs)
        batch_embeddings = outputs.last_hidden_state.mean(dim=1).numpy()
        embeddings.extend(batch_embeddings)
    return np.array(embeddings)

# Create FAISS index
def create_faiss_index(embeddings):
    dimension = embeddings.shape[1]
    nlist = min(40, len(embeddings))  # Ensure nlist <= number of training points
    quantizer = faiss.IndexFlatL2(dimension)
    index = faiss.IndexIVFFlat(quantizer, dimension, nlist, faiss.METRIC_L2)
    index.train(embeddings)
    index.add(embeddings)
    return index

# Search for similar texts
def search_similar_texts(query_embedding, index, chunks, k=3):
    distances, indices = index.search(query_embedding, k)
    similar_texts = [chunks[i] for i in indices[0]]
    return similar_texts

# Generate answer using GPT4All
def generate_answer(query, context, max_tokens=100):
    max_context_tokens = 2048 - len(tokenizer.encode(query)) - 10  # Reserve tokens for query and padding
    context_tokens = tokenizer.encode(context, max_length=max_context_tokens, truncation=True)
    truncated_context = tokenizer.decode(context_tokens, skip_special_tokens=True)
    prompt = f"Context: {truncated_context}\n\nQuestion: {query}\n\nAnswer:"
    response = gpt4all_model.generate(prompt, max_tokens=max_tokens)
    return response

# Path to PDF file
pdf_path = "/content/The_Lightning_Thief_-_Percy_Jackson_1-10.pdf"

# Read text from PDF
text = read_pdf(pdf_path)

# Split text into chunks
chunks = split_text(text, max_chunk_size=200)  # Reduced chunk size to increase the number of chunks

# Embed chunks in batches
embeddings = embed_texts_in_batches(chunks)

# Create FAISS index
index = create_faiss_index(embeddings)

# Query
query = "Who is Percy?"

# Embed query
query_embedding = embed_texts_in_batches([query])

# Search for similar texts
similar_texts = search_similar_texts(query_embedding, index, chunks, k=3)

# Generate answer
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


Answer:  Percy Jackson is a character in the book series by Rick Riordan. He is a demigod, half human and half god, with his father being Poseidon, one of the main gods in Greek mythology.



2m

# Search for similar texts
similar_texts = search_similar_texts(query_embedding, index, chunks, k=3)

# Generate answer
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
Answer:  Percy Jackson is a character in the book series by Rick Riordan. He is a demigod, half human and half god, with his father being Poseidon, one of the main gods in Greek mythology.

In [1]:
import numpy as np
import faiss
import torch
from transformers import AutoTokenizer, AutoModel
from PyPDF2 import PdfReader
from gpt4all import GPT4All
from nltk.tokenize import sent_tokenize
import nltk

# Download the required NLTK resources
nltk.download('punkt')
nltk.download('punkt_tab')

# Load GPT4All model
gpt4all_model = GPT4All("mistral-7b-openorca.Q4_0.gguf")

# Load embedding model
tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-small-en-v1.5")
embedding_model = AutoModel.from_pretrained("BAAI/bge-small-en-v1.5")

# Read PDF file
def read_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

# Split text into sentence-based chunks
def split_text(text, max_chunk_size=300):
    sentences = sent_tokenize(text)
    chunks = []
    current_chunk = ""
    for sentence in sentences:
        if len(current_chunk) + len(sentence) <= max_chunk_size:
            current_chunk += " " + sentence
        else:
            chunks.append(current_chunk.strip())
            current_chunk = sentence
    if current_chunk:
        chunks.append(current_chunk.strip())
    return chunks

# Embed texts in batches
def embed_texts_in_batches(texts, batch_size=32):
    embeddings = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tokenizer(batch, return_tensors="pt", padding=True, truncation=True, max_length=512)
        with torch.no_grad():
            outputs = embedding_model(**inputs)
        batch_embeddings = outputs.last_hidden_state.mean(dim=1).numpy()
        embeddings.extend(batch_embeddings)
    return np.array(embeddings)

# Create FAISS index
def create_faiss_index(embeddings):
    dimension = embeddings.shape[1]
    nlist = 100  # Number of clusters
    quantizer = faiss.IndexFlatL2(dimension)
    index = faiss.IndexIVFFlat(quantizer, dimension, nlist, faiss.METRIC_L2)
    index.train(embeddings)
    index.add(embeddings)
    return index

# Search for similar texts
def search_similar_texts(query_embedding, index, chunks, k=3):
    distances, indices = index.search(query_embedding, k)
    similar_texts = [chunks[i] for i in indices[0]]
    return similar_texts

# Generate answer using GPT4All
def generate_answer(query, context, max_tokens=100):
    max_context_tokens = 2048 - len(tokenizer.encode(query)) - 10  # Reserve tokens for query and padding
    context_tokens = tokenizer.encode(context, max_length=max_context_tokens, truncation=True)
    truncated_context = tokenizer.decode(context_tokens, skip_special_tokens=True)
    prompt = f"Context: {truncated_context}\n\nQuestion: {query}\n\nAnswer:"
    response = gpt4all_model.generate(prompt, max_tokens=max_tokens)
    return response

# Path to PDF file
pdf_path = "/content/The_Lightning_Thief_-_Percy_Jackson_1-10.pdf"

# Read text from PDF
text = read_pdf(pdf_path)

# Split text into chunks
chunks = split_text(text, max_chunk_size=300)

# Embed chunks in batches
embeddings = embed_texts_in_batches(chunks)

# Create FAISS index
index = create_faiss_index(embeddings)

# Query
query = "Who is Percy?"

# Embed query
query_embedding = embed_texts_in_batches([query])

# Search for similar texts
similar_texts = search_similar_texts(query_embedding, index, chunks, k=3)

# Generate answer
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


RuntimeError: Error in void faiss::Clustering::train_encoded(faiss::idx_t, const uint8_t*, const faiss::Index*, faiss::Index&, const float*) at /project/faiss/faiss/Clustering.cpp:289: Error: 'nx >= k' failed: Number of training points (38) should be at least as large as number of clusters (100)

In [1]:
import numpy as np
import faiss
import torch
from transformers import AutoTokenizer, AutoModel
from PyPDF2 import PdfReader
from gpt4all import GPT4All
from nltk.tokenize import sent_tokenize
import nltk

# Download the punkt tokenizer
nltk.download('punkt')

# Load GPT4All model
gpt4all_model = GPT4All("mistral-7b-openorca.Q4_0.gguf")

# Load embedding model
tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-small-en-v1.5")
embedding_model = AutoModel.from_pretrained("BAAI/bge-small-en-v1.5")

# Read PDF file
def read_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

# Split text into sentence-based chunks
def split_text(text, max_chunk_size=300):
    sentences = sent_tokenize(text)
    chunks = []
    current_chunk = ""
    for sentence in sentences:
        if len(current_chunk) + len(sentence) <= max_chunk_size:
            current_chunk += " " + sentence
        else:
            chunks.append(current_chunk.strip())
            current_chunk = sentence
    if current_chunk:
        chunks.append(current_chunk.strip())
    return chunks

# Embed texts in batches
def embed_texts_in_batches(texts, batch_size=32):
    embeddings = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tokenizer(batch, return_tensors="pt", padding=True, truncation=True, max_length=512)
        with torch.no_grad():
            outputs = embedding_model(**inputs)
        batch_embeddings = outputs.last_hidden_state.mean(dim=1).numpy()
        embeddings.extend(batch_embeddings)
    return np.array(embeddings)

# Create FAISS index
def create_faiss_index(embeddings):
    dimension = embeddings.shape[1]
    nlist = 100  # Number of clusters
    quantizer = faiss.IndexFlatL2(dimension)
    index = faiss.IndexIVFFlat(quantizer, dimension, nlist, faiss.METRIC_L2)
    index.train(embeddings)
    index.add(embeddings)
    return index

# Search for similar texts
def search_similar_texts(query_embedding, index, chunks, k=3):
    distances, indices = index.search(query_embedding, k)
    similar_texts = [chunks[i] for i in indices[0]]
    return similar_texts

# Generate answer using GPT4All
def generate_answer(query, context, max_tokens=100):
    max_context_tokens = 2048 - len(tokenizer.encode(query)) - 10  # Reserve tokens for query and padding
    context_tokens = tokenizer.encode(context, max_length=max_context_tokens, truncation=True)
    truncated_context = tokenizer.decode(context_tokens, skip_special_tokens=True)
    prompt = f"Context: {truncated_context}\n\nQuestion: {query}\n\nAnswer:"
    response = gpt4all_model.generate(prompt, max_tokens=max_tokens)
    return response

# Path to PDF file
pdf_path = "/content/The_Lightning_Thief_-_Percy_Jackson_1-10.pdf"

# Read text from PDF
text = read_pdf(pdf_path)

# Split text into chunks
chunks = split_text(text, max_chunk_size=300)

# Embed chunks in batches
embeddings = embed_texts_in_batches(chunks)

# Create FAISS index
index = create_faiss_index(embeddings)

# Query
query = "Who is Percy?"

# Embed query
query_embedding = embed_texts_in_batches([query])

# Search for similar texts
similar_texts = search_similar_texts(query_embedding, index, chunks, k=3)

# Generate answer
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


LookupError: 
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - '/root/nltk_data'
    - '/usr/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************


In [1]:
import numpy as np
import faiss
import torch
from transformers import AutoTokenizer, AutoModel
from PyPDF2 import PdfReader
from gpt4all import GPT4All
from nltk.tokenize import sent_tokenize

# Load GPT4All model
gpt4all_model = GPT4All("mistral-7b-openorca.Q4_0.gguf")

# Load embedding model
tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-small-en-v1.5")
embedding_model = AutoModel.from_pretrained("BAAI/bge-small-en-v1.5")

# Read PDF file
def read_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

# Split text into sentence-based chunks
def split_text(text, max_chunk_size=300):
    sentences = sent_tokenize(text)
    chunks = []
    current_chunk = ""
    for sentence in sentences:
        if len(current_chunk) + len(sentence) <= max_chunk_size:
            current_chunk += " " + sentence
        else:
            chunks.append(current_chunk.strip())
            current_chunk = sentence
    if current_chunk:
        chunks.append(current_chunk.strip())
    return chunks

# Embed texts in batches
def embed_texts_in_batches(texts, batch_size=32):
    embeddings = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tokenizer(batch, return_tensors="pt", padding=True, truncation=True, max_length=512)
        with torch.no_grad():
            outputs = embedding_model(**inputs)
        batch_embeddings = outputs.last_hidden_state.mean(dim=1).numpy()
        embeddings.extend(batch_embeddings)
    return np.array(embeddings)

# Create FAISS index
def create_faiss_index(embeddings):
    dimension = embeddings.shape[1]
    nlist = 100  # Number of clusters
    quantizer = faiss.IndexFlatL2(dimension)
    index = faiss.IndexIVFFlat(quantizer, dimension, nlist, faiss.METRIC_L2)
    index.train(embeddings)
    index.add(embeddings)
    return index

# Search for similar texts
def search_similar_texts(query_embedding, index, chunks, k=3):
    distances, indices = index.search(query_embedding, k)
    similar_texts = [chunks[i] for i in indices[0]]
    return similar_texts

# Generate answer using GPT4All
def generate_answer(query, context, max_tokens=100):
    max_context_tokens = 2048 - len(tokenizer.encode(query)) - 10  # Reserve tokens for query and padding
    context_tokens = tokenizer.encode(context, max_length=max_context_tokens, truncation=True)
    truncated_context = tokenizer.decode(context_tokens, skip_special_tokens=True)
    prompt = f"Context: {truncated_context}\n\nQuestion: {query}\n\nAnswer:"
    response = gpt4all_model.generate(prompt, max_tokens=max_tokens)
    return response

# Path to PDF file
pdf_path = "/content/The_Lightning_Thief_-_Percy_Jackson_1-10.pdf"

# Read text from PDF
text = read_pdf(pdf_path)

# Split text into chunks
chunks = split_text(text, max_chunk_size=300)

# Embed chunks in batches
embeddings = embed_texts_in_batches(chunks)

# Create FAISS index
index = create_faiss_index(embeddings)

# Query
query = "Who is Percy?"

# Embed query
query_embedding = embed_texts_in_batches([query])

# Search for similar texts
similar_texts = search_similar_texts(query_embedding, index, chunks, k=3)

# Generate answer
context = " ".join(similar_texts)
answer = generate_answer(query, context)

print("Answer:", answer)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


LookupError: 
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - '/root/nltk_data'
    - '/usr/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************
